### DISCUSSION TITLE
Dropping requirement of implicit typing for MPI-projects

### DISCUSSION THREAD

**Discussion by aradi** [fortran-lang/fpm#1035] [Ideas]
Currently, fpm requires implicit typing and implicit externals on projects with MPI dependency. While I see its necessity with older MPI-frameworks and the f77 and the (often incorrectly implemented) f90 interface, it enforces a bad and dangerous practice. As new Fortran practitioners  are likely to use fpm as their first build system (it is really cool to have fpm!), they might get the wrong impression about the language.

My suggestions would be not to require any implicit features for any MPI-dependencies. The general advice for new projects should be anyway to use the mpi_f08 interface. According to my experience, this is implemented correct enough  in current OpenMPI, MPICH and IntelMPI, and works in each of them without any implicit tricks. One might add some  additional notes in the documentation about the necessity of implicit typing/externals for older MPI-frameworks and the older MPI-interfaces, but the default setting should support "best practice".

**Comment by perazz** [comment]
I also strongly support modern coding practices and would like to see this become reality. 
Currently, dropping the implicit typing option would mean dropping MS Windows support altogether for non-Intel compilers.
The reason is that OpenMPI and MPICH are not supported. In contrast, MS-MPI (that is based on MPICH 1.4) does not support modern Fortran modules (`use mpi`, `use mpi_f08`). What options do we have? 

- Lift the requirement except on Windows+MS-MPI --> drop `mpif.h` support on Unix/macOS only. 
- Lift the requirement altogether -> Drop support for Windows+MS-MPI
- Do not enforce implicit typing features on the MPI metapackage --> let the user expertise create compilable code

**Reply by ivan-pi** [reply]
Is there any chance to work with Microsoft to provide a Fortran module?

**Reply by awvwgk** [reply]
The implementation of MS-MPI is available as open source at https://github.com/microsoft/Microsoft-MPI. I don't know who is owning the repository at the moment or whether it is still active.

**Comment by aradi** [comment]
@perazz Thanks for explaining, I was not aware of the issue with Windows :roll_eyes: .

But do we really need implicit typing, or would a somehow propagated `-fallow-argument-mismatch` like argument suffice? Looking quickly at `mpif.h` in MPICH, I don't see anything which would require implicit typing. Implicit externals are indeed needed, I agree.
